base/compiler/tfuncs.jl

```diff
@@ -826,17 +826,18 @@ function getfield_tfunc(@nospecialize(s00), @nospecialize(name))
             end
         end
         s = typeof(sv)
-    elseif isa(s, PartialStruct)
+    elseif isa(s00, PartialStruct)
+        s = widenconst(s00)
+        sty = unwrap_unionall(s)::DataType
```

Comment:
Just to make sure: is there any possibility where `s::DataType` doesn't hold ?
This check is good to have though, we may want to form `PartialStruct` for abstract types other than tuples in the future.

Reply:
Currently, no. That is why this PR has sat for so long untouched. But turns out to be fairly straightforward here to add (though I think the necessary type lattice changes may currently be difficult).